### PR TITLE
Credentials: retry credentials on SDK credential errors

### PR DESCRIPTION
**Pull request by gsweetwood**
## Problem
Credentials can become expired. All SDK network calls will fail and
the toolkit only logs the error.
## Solution
Add a request listener to the SDK clients that can catch
expired credential errors. Immediately retry the current profile. Credentials are static, so if they were refreshed, this will pull the new credentials. If that is unsuccessful, give the option to retry or to change profiles.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/requestListeners.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import { localize } from 'vscode-nls'
+import { CredentialsStore } from '../credentials/credentialsStore'
+import { LoginManager } from '../credentials/loginManager'
+import { ServiceOptions } from './awsClientBuilder'
+import { CredentialsProfileMru } from './credentials/credentialsProfileMru'
+import globals from './extensionGlobals'
+
+export const CREDENTIAL_ERROR_REQUEST_LISTENER: ServiceOptions = {
+    onRequestSetup: [
+        req => {
```
Comment (JadenSimon) [inline src/shared/requestListeners.ts:16]: The `onRequestSetup` callback can be fired multiple times, but we have no guard against this. An easy solution is to just add a flag that is set on the first expiration error. Then the listener can return early if the flag is already set.

File: src/shared/requestListeners.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import * as vscode from 'vscode'
+import { localize } from 'vscode-nls'
+import { CredentialsStore } from '../credentials/credentialsStore'
+import { LoginManager } from '../credentials/loginManager'
+import { ServiceOptions } from './awsClientBuilder'
+import { CredentialsProfileMru } from './credentials/credentialsProfileMru'
+import globals from './extensionGlobals'
+
+export const CREDENTIAL_ERROR_REQUEST_LISTENER: ServiceOptions = {
```
Comment (JadenSimon) [inline src/shared/requestListeners.ts:14]: If we're calling this file `requestListeners`, then it'd make more sense to export things with the `RequestListener` shape, not `ServiceOptions`

File: src/shared/requestListeners.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            req.on('error', async err => {
+                if (
+                    err.name.includes('ExpiredToken') ||
+                    err.message.includes('The security token included in the request is expired')
+                ) {
+                    const loginManager = new LoginManager(globals.awsContext, new CredentialsStore())
+                    let retryLoginResult = await loginManager.retryLogin()
+                    const profileName = new CredentialsProfileMru(globals.context).getMruList()[0]
+                    if (retryLoginResult) {
+                        vscode.window.showInformationMessage(
```
[the remark this answers is not in the document]
Comment (gsweetwood) [reply src/shared/requestListeners.ts:26]: My one concern here is that their SDK call will fail. As a user, I wouldn't know that I can just rerun it and it should work. The status bar text could work though. It's more subtle at least.

[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/requestListeners.ts:26]: In the current implementation, by the time this code is executed the request will fail regardless. The handlers do not block the request error from returning to the original caller. So right now we'd show two notifications: a generic request failure and either a "success" on refresh or another failure.

A cohesive implementation needs to communicate with other logic somehow. Possible options:
* rewrite the request error to something Toolkit specific
* mutate either the request object or request error to attach Toolkit data
* use a shared reference injected at request time

**Pull request closed without merging**